mlflow/onnx.py

```diff
@@ -189,27 +189,36 @@ def predict(self, dataframe):
         # client and the scoring server can introduce 64-bit floats. This is being tracked in
         # https://github.com/mlflow/mlflow/issues/1286. Meanwhile, we explicitly cast the input to
         # 32-bit floats when needed. TODO: Remove explicit casting when issue #1286 is fixed.
-        if len(self.inputs) > 1:
-            cols = [name for (name, type) in self.inputs if type == "tensor(float)"]
-        else:
-            cols = dataframe.columns if self.inputs[0][1] == "tensor(float)" else []
-
-        dataframe = _OnnxModelWrapper._cast_float64_to_float32(dataframe, cols)
-        if len(self.inputs) > 1:
-            feed_dict = {name: dataframe[name].values for (name, _) in self.inputs}
-        else:
-            feed_dict = {self.inputs[0][0]: dataframe.values}
-        predicted = self.rt.run(self.output_names, feed_dict)
 
-        def format_output(data):
-            # Output can be list and it should be converted to a numpy array
-            # https://github.com/mlflow/mlflow/issues/2499
-            data = np.asarray(data)
-            return data.reshape(-1)
+        number_inputs = len(self.inputs)
+        number_columns = len(dataframe.columns)
+        if number_inputs == number_columns:
+            column_data = {}
+            for inp in self.inputs:
+                values = np.stack(dataframe[inp.name].values)
```

Comment:
In what cases does `dataframe[inp.name].values` differ from `np.stack(dataframe[inp.name].values)`? Is this stacking necessary?

Reply:
This is an important difference, partially documented in the example of #3229. I will give a more concrete example here of a pandas dataframe with one column of shape (2,2,1)

`df = pd.DataFrame(data={'col1' : [[[1], [2]], [[3], [4]]]})`
```
df['col1'].values.shape
(2,)
```
```
np.stack(df['col1'].values)
array([[[1],
        [2]],

       [[3],
        [4]]])
>>> np.stack(df['col1'].values).shape
(2, 2, 1)
```

Without the np.stack, the object corresponding to dataframe[col].values is always of rank 2 no matter what its actual shape is.
By applying the np.stack we can bring the data in the correct shape before we feed it to the model. Without the np.stack the model input would always be of rank 2 even if the model expects higher dimensional input. I will add comments about that.